internal/driver/driver.go

```diff
@@ -0,0 +1,790 @@
+	s.lc.Debugf("Device %s is removed", deviceName)
+	s.mu.Lock()
+	s.s7Clients[deviceName] = nil
+	s.mu.Unlock()
+	return nil
+}
+
+func (s *Driver) ValidateDevice(device models.Device) error {
+	s.lc.Debugf("Validating device: %s", device.Name)
+
+	protocols := device.Protocols
+	pp := protocols[Protocol]
+	var errt error
+
+	if pp == nil {
+		errt = fmt.Errorf("%s not found in protocols for device %s", Protocol, device.Name)
+		s.lc.Error(errt.Error())
+		return errt
+	}
+
+	_, errt = cast.ToStringE(pp["Host"])
+	if errt != nil {
+		s.lc.Errorf("Host not found or not a string in Protocol, error: %s", errt)
+		return errt
+	}
+	_, errt = cast.ToIntE(pp["Port"])
+	if errt != nil {
+		s.lc.Errorf("Port not found or not an integer in Protocol, error: %s", errt)
+		return errt
+	}
+	_, errt = cast.ToIntE(pp["Rack"])
+	if errt != nil {
+		s.lc.Errorf("Rack not found or not an integer in Protocol, error: %s", errt)
+		return errt
+	}
+	_, errt = cast.ToIntE(pp["Slot"])
+	if errt != nil {
+		s.lc.Errorf("Slot not found or not an integer in Protocol, error: %s", errt)
+		return errt
+	}
+	_, errt = cast.ToIntE(pp["Timeout"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		pp["Timeout"] = 30
+	}
+	_, errt = cast.ToIntE(pp["IdleTimeout"])
+	if errt != nil {
+		s.lc.Errorf("IdleTimeout not found or not an ingeger in Protocol, USE DEFAULT 30s, error: %s", errt)
+		pp["IdleTimeout"] = 30
+	}
+
+	return nil
+}
+
+// Create S7Client by 'Device' definition
+func (s *Driver) NewS7Client(deviceName string, protocol map[string]models.ProtocolProperties) *S7Client {
+
+	pp := protocol[Protocol]
+
+	host, _ := cast.ToStringE(pp["Host"])
+	port, _ := cast.ToStringE(pp["Port"])
+	rack, _ := cast.ToIntE(pp["Rack"])
+	slot, _ := cast.ToIntE(pp["Slot"])
+	timeout, _ := cast.ToIntE(pp["Timeout"])
+	idletimeout, _ := cast.ToIntE(pp["IdleTimeout"])
+
+	// create handler: PLC tcp client
+	handler := gos7.NewTCPClientHandler(host+":"+port, rack, slot)
+	if handler == nil {
+		s.lc.Errorf("Cant not create NewTCPClientHandler: %s", handler)
+		return nil
+	}
+	s.lc.Debugf("New TCP Client: %s", handler)
+
+	// handler connect timeout from 'Timeout'
+	handler.Timeout = time.Duration(timeout) * time.Second
+
+	// handler connect idle timeout from 'IdleTimeout'
+	handler.IdleTimeout = time.Duration(idletimeout) * time.Second
+
+	// connect to S7
+	err := handler.Connect()
+	if err != nil {
+		s.lc.Errorf("Can't handler S7 Connect: %s, error: %s", deviceName, err)
+		// return nil
+	}
+
+	s7client := gos7.NewClient(handler)
+	client := &S7Client{
+		DeviceName: deviceName,
+		Client:     s7client,
+	}
+	return client
+
+}
+
+// Get S7Client by 'DeviceName'
+func (s *Driver) getS7Client(deviceName string, protocols map[string]models.ProtocolProperties) *S7Client {
+	s.mu.Lock()
+	s7Client := s.s7Clients[deviceName]
+	s.mu.Unlock()
+
+	if s7Client == nil {
+		s.lc.Warnf("Can not get S7CLient from: %s", deviceName)
```

Comment:
Sory, this is should be a not found message like:
```suggestion
		s.lc.Warnf("S7CLient for device %s not found. Creating it...", deviceName)
```